### ISSUE TITLE
Event and Refactoring Request for FinalizePosting in codeunits 80 and 90 

### ISSUE THREAD

**Issue by NKarolak** [Microsoft/AL#1338]
Currently, **function FinalizePosting in codeunits 80 and 90** will delete a completely invoiced order. However, partners sometimes require to keep the invoiced order anyway. Therefore, please add a boolean variable (e.g. KeepOrder) to FinalizePosting and add a handled pattern event. Suggestion:

![001](https://user-images.githubusercontent.com/34504100/34554373-789b2d56-f124-11e7-9d72-5e0cd6bc2797.JPG)

However, I can imgine that other partners probably would need more parameters for OnBeforeFinalizePosting, maybe even the EverythingInvoiced variable?

**Comment by AlexanderYakunin** [comment]
We already have ArchiveManagement.AutoArchiveSalesDocument(SalesHeader); for completely invoiced orders. You just need to set Archive Orders in Sales Setup. Keeping fully posted orders in Sales Header/Line tables decrease performance as these tables are heavily used in UI. After archiving you can always view and restore your order back, if needed. Based on this do you really need to keep posted sales orders?

**Comment by NKarolak** [comment]
I understand your concerns, but the suggestion does not fit our needs, as we only want to postpone the deletion:
We do not intend to keep all invoiced documents, but only particular ones, based on custom fields.
The process is: After complete invoicing, the user checks and updates a few custom fields for an EDI message to generate. Once this process is finished, the user will delete the Order manually and archive it - but not earlier.
